task-maker-async/src/store.rs

```diff
@@ -43,58 +82,591 @@ pub enum FileReadingOutcome {
     /// to it. It is an error to refresh a lease of a non-existent input.
-    async fn refresh_fileset_lease(handle: FileSetHandle) -> Result<(), ()>;
+    async fn refresh_fileset_lease(handle: FileSetHandle) -> Result<(), Error>;
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum FileSetKind {
+    Computation,
+    InputFile,
+}
+
+type FileContents = Vec<u8>;
+
+#[derive(Debug)]
+struct FileSet {
+    files: HashMap<FileSetFile, FileContents>,
+    kind: FileSetKind,
+    finalized: bool,
 }
 
+type FileSetVariants = HashMap<HashData, FileSet>;
+
+#[derive(Debug)]
+struct FileSetHandleInfo {
+    outer_hash: HashData,
+    inner_hash: HashData,
+    expiration: Instant,
+    mode: HandleMode,
+    file_handles: HashMap<FileHandleId, FileSetFile>,
+    next_handle: FileHandleId,
+}
+
+struct StoreServiceImpl {
+    filesets: HashMap<HashData, FileSetVariants>,
+    file_set_handles: HashMap<FileSetHandleId, FileSetHandleInfo>,
+    next_handle: FileSetHandleId,
+}
+
+impl StoreServiceImpl {
+    fn new() -> Self {
+        StoreServiceImpl {
+            filesets: HashMap::new(),
+            file_set_handles: HashMap::new(),
+            next_handle: 0,
+        }
+    }
+
+    fn new_handle(&mut self, hash: ComputationHash, mode: HandleMode) -> FileSetHandle {
+        let handle = self.next_handle;
+        self.next_handle += 1;
+        let ComputationHash(outer_hash, inner_hash) = hash;
+        let handle_info = FileSetHandleInfo {
+            outer_hash,
+            inner_hash,
+            expiration: Instant::now() + LEASE_LENGTH,
+            mode,
+            file_handles: HashMap::new(),
+            next_handle: 0,
+        };
+        self.file_set_handles.insert(handle, handle_info);
+        FileSetHandle { id: handle, mode }
+    }
+
+    fn create_or_open_fileset(
+        &mut self,
+        hash: ComputationHash,
+        kind: FileSetKind,
+    ) -> Result<FileSetHandle, Error> {
+        let outer_comp = self.filesets.entry(hash.0.clone()).or_default();
+        if let Some(file_set) = outer_comp.get(&hash.1) {
+            if file_set.kind != kind {
+                return Err(Error::HashCollision(hash));
+            }
+            Ok(self.new_handle(hash, HandleMode::Read))
+        } else {
+            outer_comp.insert(
+                hash.1.clone(),
+                FileSet {
+                    files: HashMap::new(),
+                    kind,
+                    finalized: false,
+                },
+            );
+            Ok(self.new_handle(hash, HandleMode::Write))
+        }
+    }
+
+    fn clear_expired_leases(&mut self) {
+        let time = Instant::now();
+        for info in self.file_set_handles.values() {
+            if info.expiration < time && info.mode == HandleMode::Write {
+                self.filesets
+                    .get_mut(&info.outer_hash)
+                    .unwrap()
+                    .remove(&info.inner_hash);
```

Comment:
Does this remove all the filesets not pointed by any handle?

Reply:
Now I see, this applies only to write handles. So the file is removed if it would be truncated (and not finalized). A comment may be ideal here to explain this.